gravitational/teleport
# Support routing to Kubernetes clusters by request path

This implements path-based routing for Kubernetes clusters as described by [RFD0185]. A new prefixed path handler is added that accepts base64-encoded Teleport and Kubernetes cluster names. The request is routed to the destination Teleport cluster using these parameters instead of those embedded in the session TLS identity, and then the preexisting handlers check authorization and complete the request as usual.

This removes the need for certificates to be issued per Kubernetes cluster: so long as the incoming identity is granted access to the cluster via its roles, access can succeed, and no `KubernetesCluster` attribute or cert usage restrictions are needed.

Fixes #40405

No changelog for this PR, will be added in #50898

[RFD0185]: https://github.com/gravitational/teleport/pull/47436

---

### Testing steps

1. Ensure a Teleport proxy is running this branch with one or more k8s clusters attached. The Kubernetes node does not need to be running this branch, only the proxy.
2. Build `tbot` against #50898
3. Create a new bot with appropriate permissions to access your kube node(s), e.g. `tctl bots add foo --roles=kube-access`
4. Start `tbot` with the new `kubernetes/v2` service:
   ```
   $ tbot start kubernetes/v2 --destination=./tbot-user --storage=./tbot-data --token=$token --join-method=token --kubernetes-cluster-name default --proxy-server=example.teleport.sh:443
   ```

   The `--kubernetes-cluster-name` flag can be repeated. Alternatively (or additionally), try a label matcher:
   ```
   $ tbot start kubernetes/v2 --destination=./tbot-user --storage=./tbot-data --token=$token --join-method=token --kubernetes-cluster-labels foo=bar --proxy-server=example.teleport.sh:443
   ```
6. Examine `./tbot-user/kubeconfig.yaml`. You should see one `cluster:` and one `context:` entry per matched cluster. The `server:` field should include a new URL suffix, `/v1/teleport/<teleportCluster>/<kubeCluster>`
7. Try `KUBECONFIG=./tbot-user/kubeconfig.yaml kubectl get pods`. Also consider testing `kubectl exec`, port forward, etc.
